repl: Add initial notebook execution + KernelSession abstraction

MostlyKIGuess opened the pull request:
Coming from discussion on #25936 and https://github.com/zed-industries/zed/pull/19756 . 
This PR introduces the KernelSession abstraction and adds basic notebook cell execution inside NotebookEditor.

The following provides a base starter for the development on Notebooks.

What this PR includes today:

Release Notes:

- KernelSession trait extracted. Both REPL and NotebookEditor now share the same routing mechanism.
- NotebookEditor can launch kernels, execute code cells, and display outputs.
- Basic cell operations: run current cell, run all, move up/down, add code/markdown blocks. Keybindings follow Jupyter defaults.


Next Steps:

- [x] Editing support for markdown and code blocks.
- Buffer integration instead of temporary cell text.
- [x] Proper focus behavior when executing or adding cells.
- Kernel control UI.

A little far fetched steps:

- Vim Support
- Cell Handling Improvement and other convenient features and design from other editors
- Ability to have better parsing for AI Support. 


I have attached a video of showcasing some of the features:

https://github.com/user-attachments/assets/37e6f3e5-2022-45f0-a73d-2dd01ebc2932

MostlyKIGuess committed:
repl: Extract KernelSession trait for message routing

- Introduces a KernelSession trait that abstracts kernel message routing, allowing both Session (REPL) and NotebookEditor to use the same kernel infrastructure.

MostlyKIGuess committed:
repl: Add notebook cell execution support

- Add kernel launching in NotebookEditor
- Implement cell execution with output handling
- Add CodeCell constructors and message handling
- Subscribe to CellEvent::Run for button clicks
- Note: Markdown Handling is still yet to be done, focus on editing is also left and no change has been made to save the buffer.

MostlyKIGuess committed:
keymaps: Add NotebookEditor keybindings

- shift-enter/ctrl-enter: Run current cell
- ctrl-shift-enter: Run all cells
- alt-up/alt-down: Move cell up/down
- ctrl-m: Add code block
- ctrl-shift-m: Add markdown block

MostlyKIGuess committed:
Merge branch 'zed-industries:main' into jupyter

MostlyKIGuess committed:
fix: fmt and removed unnecessary clone

GustavoWidman commented:
<img width="736" height="490" alt="image" src="https://github.com/user-attachments/assets/e3e31ac7-53fb-4795-a309-43cf15f7e4d8" />

MostlyKIGuess commented:
Hey @osiewicz would you like to contact over discord? I was able to make the markdown editing as well as the focus running work, so perhaps this might get bigger. It would be nice to know your expectations of code

MostlyKIGuess committed:
feat: add editable Markdown cells + focus fix

- inline Editor for MarkdownCell with editing state and reparse task that runs when editing finishes.
-  Preview shows rendered
markdown; clicking or focusing enters edit mode, and blur/Run finishes
editing and triggers reparse.
- fix: NotebookEditor subscribes to editor focus, fixing the issue of focused cell running!

MostlyKIGuess committed:
Merge remote-tracking branch 'upstream/main' into jupyter

lhutyra commented:
> Updated the comment on what next steps are achieved!

I really apprieciate this PR, and hope it will be merged soon, at least as a first step to larger notebook support. It was paused for months

MostlyKIGuess committed:
fix: notebook shortcut preference

lhutyra commented:
> What do you think the next step should be?

I will build this branch, and will do some tests with my notebooks

MostlyKIGuess committed:
feat: add cell language badge

- fix: re-add the enter -> newline in keymaps

Pd-ch commented:
Consider these features? Display the execution time of cells (low priority), and be able to use % or ! to execute shell lines like in VSCode？

MostlyKIGuess commented:
@Pd-ch , time counters :D 

<img width="569" height="753" alt="image" src="https://github.com/user-attachments/assets/a2bfbc81-15e9-4dcd-8ff0-7b1160bece6e" />
<img width="767" height="693" alt="image" src="https://github.com/user-attachments/assets/92ece9bf-bd8c-484c-9796-4d4bc9744d46" />

MostlyKIGuess commented:
I am trying to get the Kernel UI working, should be able to push this by the weekend, along with life enhancements such as, kernel status running, idle etc and to add interrupt, restart options

MostlyKIGuess committed:
Add kernel controls and execution timing

rgbkrk commented:
On first look, this is great. You built right on top of the foundations we laid out. Since this is still behind a feature flag anyhow, I'll recommend to Zed maintainers that they bring it for you and others to keep contributing to. Looking forward to using this!

MostlyKIGuess committed:
Merge remote-tracking branch 'upstream/main' into jupyter

MostlyKIGuess committed:
fix: duplicate fn merge

MostlyKIGuess committed:
merge to upstream zed

lict99 commented:
Thank you for this great PR! I really appreciate the work. I'm just wondering will this also support [interactive notebooks](https://code.visualstudio.com/docs/python/jupyter-support-py) similar to the experience in VS Code?

Pd-ch commented:
Although notebooks are listed last in the Zed 1.0 roadmap, I still hope that basic support can be added to Zed as soon as possible, which would greatly meet the needs of many people, especially those who are already tired of VSCode. Finally, thank you and the Zed developers for your efforts.

lhutyra commented:
> Although notebooks are listed last in the Zed 1.0 roadmap, I still hope that basic support can be added to Zed as soon as possible, which would greatly meet the needs of many people, especially those who are already tired of VSCode. Finally, thank you and the Zed developers for your efforts.

Yes, thats true, it's only reason, why I still need to keep vs code.

MostlyKIGuess committed:
Add notebook save/load

MostlyKIGuess committed:
Fix window.focus() calls to use updated GPUI API

MostlyKIGuess committed:
Merge remote-tracking branch 'upstream/main' into jupyter

MostlyKIGuess committed:
Merge remote-tracking branch 'upstream/main' into jupyter

MostlyKIGuess committed:
Fix window.focus() calls to use updated GPUI API

MostlyKIGuess committed:
Merge remote-tracking branch 'upstream/main' into jupyter

MostlyKIGuess committed:
fix: import KernelSession after rebase

MostlyKIGuess commented:
> I'm trying to run this now and it's not building locally.
> 
> ```
> error[E0061]: this method takes 2 arguments but 1 argument was supplied
>     --> crates/repl/src/notebook/cell.rs:613:44
>      |
>  613 | ...                   window.focus(&this.editor.focus_handle(cx));
>      |                              ^^^^^------------------------------- argument #2 of type `&mut gpui::App` is missing
>      |
> note: method defined here
>     --> /Users/kylekelley/code/src/github.com/zed-industries/zed/crates/gpui/src/window.rs:1487:12
>      |
> 1487 |     pub fn focus(&mut self, handle: &FocusHandle, cx: &mut App) {
>      |            ^^^^^
> help: provide the argument
>      |
>  613 |                                     window.focus(&this.editor.focus_handle(cx), /* &mut gpui::App */);
>      |                                                                               ++++++++++++++++++++++
> ```
> 
> You might want to rebase and make sure all is smooth sailing after the rebase.

Fixed and merged. This will build now.

MostlyKIGuess committed:
fix: style error

maxdeviant commented on crates/repl/src/notebook/notebook_ui.rs:
```diff
@@ -128,16 +236,273 @@ impl NotebookEditor {
+        window: &mut Window,
+        cx: &mut Context<Self>,
+    ) {
+        // shutdown the current kernel if running
```
A lot of these comments are unnecessary, as they are just repeating what the code is doing.

MostlyKIGuess committed:
style: remove obvious comments

MostlyKIGuess committed:
Merge remote-tracking branch 'upstream/main' into jupyter

MostlyKIGuess committed:
Merge remote-tracking branch 'upstream/main' into jupyter

MostlyKIGuess committed:
Fix compilation error in Markdown initialization

MostlyKIGuess committed:
Merge remote-tracking branch 'upstream/main' into jupyter

MostlyKIGuess committed:
added Mac and Windows keybinds and removed extra comments

MostlyKIGuess committed:
Merge remote-tracking branch 'upstream/main' into jupyter

Pd-ch commented:
@kndbvortex 
First,
```
git clone https://github.com/MostlyKIGuess/zed.git -b jupyter
```
Then follow the [instructions](https://github.com/MostlyKIGuess/zed/tree/jupyter#developing-zed) to compile and install.This is not an extension.Remember, you need to set the environment variables:
```
LOCAL_NOTEBOOK_DEV=1
```
